tools/git-change-exec/main.go

```diff
@@ -0,0 +1,307 @@
+	refs = append(refs, gce.retrieveLtsRefs()...)
+
+	for _, ref := range refs {
+		commit, err := gce.g.CommitObject(ref.Hash())
+		if err != nil {
+			log.Printf("retrieve commit object from ref %v failed: %v", ref, err)
+			continue
+		}
+		commonBase = append(commonBase, commit)
+		addBase, err := branchHead.MergeBase(commit)
+		if err != nil {
+			log.Printf("finding merge base failed: %v", err)
+		}
+		commonBase = append(commonBase, addBase...)
+	}
+	return commonBase
+}
+
+func (gce *gitChangeExec) retrieveLtsRefs() []*plumbing.Reference {
+	ret := []*plumbing.Reference{}
+
+	refs, err := gce.g.References()
+	if err != nil {
+		log.Printf("retrieve refs failed: %v", err)
+	}
+
+	err = refs.ForEach(func(r *plumbing.Reference) error {
+		// f.e. refs/remotes/origin/10.4-stable
+		if strings.HasPrefix(r.Name().String(), "refs/remotes/origin") &&
+			strings.HasSuffix(r.Name().String(), "-stable") {
+			ret = append(ret, r)
+		}
+		return nil
+	})
+
+	if err != nil {
+		log.Printf("iterating over refs failed: %v", err)
+	}
+
+	return ret
+}
+
+func (gce *gitChangeExec) retrieveMasterRef() []*plumbing.Reference {
+	masterRefs := []*plumbing.Reference{}
+
+	for _, nameOfMaster := range []string{
+		"refs/heads/master",
+		"refs/remotes/origin/master",
+	} {
+		var err error
+
+		masterRef, err := gce.g.Reference(plumbing.ReferenceName(nameOfMaster), true)
+		if err == nil {
+			masterRefs = append(masterRefs, masterRef)
+		}
+	}
+	return masterRefs
+}
+
+func (gce *gitChangeExec) addActionByPath(path string) {
+	if path == "" {
+		return
+	}
+	fp := filepath.Join(gce.gitPath, path)
+	_, visited := gce.visitedPaths[fp]
+	if visited {
+		return
+	}
+	gce.visitedPaths[fp] = struct{}{}
+	for _, a := range gce.actionsToCheck {
+		if a.match(path) {
+			gce.actionDos[a] = struct{}{}
+		}
+	}
+}
+
+func (gce *gitChangeExec) runActionDos() {
+	failed := false
+	for _, a := range gce.actionsToCheck {
+		_, found := gce.actionDos[a]
+		if !found {
+			continue
+		}
+		var err error
+		if !dryRun {
+			log.Printf("--- running %s ...", id(a))
+			err = a.do()
+			log.Printf("--- running %s done", id(a))
+		} else {
+			log.Printf("would run %s, but running dry ...", id(a))
+		}
+		if err != nil {
+			log.Printf("%s failed with: %v", id(a), err)
+			failed = true
+		}
+	}
+
+	if failed {
+		os.Exit(1)
+	}
+}
+
+func (gce *gitChangeExec) collectDirtyGitTree() {
+	ignoredStatusCodes := map[git.StatusCode]struct{}{
+		git.Unmodified: struct{}{},
+		git.Untracked:  struct{}{},
+	}
+	worktree, err := gce.g.Worktree()
+	if err != nil {
+		log.Fatalf("getting current worktree: %v", err)
+	}
+
+	stats, err := worktree.Status()
+	if err != nil {
+		log.Fatalf("getting current worktree status: %v", err)
+	}
+
+	for file, gitSt := range stats {
+		_, foundStaging := ignoredStatusCodes[gitSt.Staging]
+		_, foundWorkTree := ignoredStatusCodes[gitSt.Worktree]
+
+		if foundStaging && foundWorkTree {
+			continue
+		}
+
+		fp := filepath.Join(gce.gitPath, file)
```

Comment:
Shouldn't we ignore the removed files?

Reply:
it does an `os.Stat` here: https://github.com/lf-edge/eve/pull/4202/files#diff-711d61f4f130257efb2bca48eebbdc4d829cf9f3f687f00014abe0c9aafc2ccfR297
and ignores the file if it does not exist.

Reply:
Ah, sorry. I missed the lines

Reply:
I, too :rofl: 
when I tried to find out ...